### ISSUE TITLE
Don't store unused tensors in memory when loading a safetensors file

### ISSUE THREAD

**Issue by Nerogar** [huggingface/safetensors#542]
### Feature request

It would be great if we could release unused tensors from memory when they are no longer needed.

### Motivation

In [OneTrainer](https://github.com/Nerogar/OneTrainer) I added a feature to convert the data type of a model while it's loading. For example, you can load from a safetensors file with fp32 weights, but convert them to fp8 to reduce ram/vram usage. Weights are loaded one by one, and directly converted to their target data type. This is done by using the lazy loading of safetensors, similar to this example:

```python
def load():
    filename = "path/to/fp32.safetensors"
    state_dict = load_file(filename)
    return [tensor for tensor in state_dict.values()]

if __name__ == '__main__':
    states = load()

    for i in range(len(states)):
        states[i] = states[i].to(dtype=torch.float8_e4m3fn)
```

Now the problem is that the entire fp32.safetensors file is kept in memory until every tensor is converted. For a 10GB file, this means the memory usage spikes to 12.5GB (10GB for the file, and 2.5GB for the fp8 copy), then drops down to 2.5GB

It would be great if there was some kind of release mechanism, either manual or automatic, that can be used to remove that 10GB spike

### Your contribution

-

**Comment by deepbeepmeep** [comment]
I completely agree. It is a big inconvenient of safetensors especially for low mem configs .It is too bad since safetensors represent a great advancement.

**Comment by Narsil** [comment]
```python
with safe_open(filename, device=xxx") as f:
   tensor = f.get_tensor("tensor_name")
   delete tensor
```

This is in the docs and the readme: https://huggingface.co/docs/safetensors/index#load-tensors

**Comment by Nerogar** [comment]
I don't think this actually solves the problem. I just tried this. "H:/stable-diffusion/models/ckpt-base/flux1-dev.safetensors" is the flux dev model in bf16 format, downloaded from https://huggingface.co/black-forest-labs/FLUX.1-dev/tree/main

```python
import time

import torch
from safetensors import safe_open

filename = "H:/stable-diffusion/models/ckpt-base/flux1-dev.safetensors"

if __name__ == '__main__':
    states = []
    with safe_open(filename, framework="pt") as f:
        for k in f.keys():
            tensor = f.get_tensor(k)
            states.append(tensor.to(dtype=torch.float8_e4m3fn))
            del tensor

    time.sleep(5)
```

And here is a task manager screenshot during the script execution.
![image](https://github.com/user-attachments/assets/6dc7df50-c651-445b-8d52-b6f98f18471f)

As you can see, there is a spike, then a 5s plateau which corresponds to the time.sleep(5) call. I would expect a much lower peak memory usage, just slightly higher than the 5s plateau at the end.

I also tried this to completely remove any access to other values of the loaded file during the tensor load phase. But the result is exactly the same:
```python
import time

import torch
from safetensors import safe_open

filename = "H:/stable-diffusion/models/ckpt-base/flux1-dev.safetensors"

if __name__ == '__main__':
    states = []
    keys = []
    with safe_open(filename, framework="pt") as f:
        keys.extend(f.keys())

    with safe_open(filename, framework="pt") as f:
        for k in keys:
            tensor = f.get_tensor(k)
            states.append(tensor.to(dtype=torch.float8_e4m3fn))
            del tensor

    time.sleep(5)
```

**Comment by Nerogar** [comment]
@Narsil could you take another look at this? Maybe I'm doing something wrong, but I couldn't solve my problem with the info you provided.
